### PR TITLE
Optimize baseline and special case performance for kSmallest

### PR DESCRIPTION
**Pull request by Barbarrosa**
Improve the baseline and special case performance of `kSmallest`.

Before:

```
starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 26.71 ops/sec ±1.26% (48 runs sampled)

starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 2.75 ops/sec ±1.76% (11 runs sampled)
```

After:

```
starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 36.96 ops/sec ±1.03% (64 runs sampled)

starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 3.66 ops/sec ±1.26% (14 runs sampled)
```

### CODE REVIEW

File: FastPriorityQueue.js [hunk 1]
```diff
@@ -287,13 +287,22 @@ FastPriorityQueue.prototype.forEach = function(callback) {
 FastPriorityQueue.prototype.kSmallest = function(k) {
   if (this.size == 0) return [];
   k = Math.min(this.size, k);
-  var fpq = new FastPriorityQueue(this.compare);
-  const newSize = Math.min((k > 0 ? Math.pow(2, k - 1) : 0) + 1, this.size);
+  const newSize = Math.min(this.size, (1 << (k - 1)) + 1);
```
Comment (lemire) [inline FastPriorityQueue.js:290]: We had a guard against k == 0 which this PR drops. When k = 0, we get (1<<(-1))+1 which is a silly value.

File: FastPriorityQueue.js [hunk 2]
```diff
@@ -287,13 +287,22 @@ FastPriorityQueue.prototype.forEach = function(callback) {
 FastPriorityQueue.prototype.kSmallest = function(k) {
   if (this.size == 0) return [];
```
Comment (lemire) [inline FastPriorityQueue.js]:
```suggestion
  if ((this.size == 0) || (k<=0)) return [];
```

File: FastPriorityQueue.js (hunk 1 above)
Comment (lemire) [reply FastPriorityQueue.js:290]: If we guard against k == 0 previously, then this code will look safer (k==1 is fine, k < 1 is not).

File: FastPriorityQueue.js [hunk 3]
```diff
@@ -287,13 +287,22 @@ FastPriorityQueue.prototype.forEach = function(callback) {
...
   k = Math.min(this.size, k);
-  var fpq = new FastPriorityQueue(this.compare);
-  const newSize = Math.min((k > 0 ? Math.pow(2, k - 1) : 0) + 1, this.size);
+  const newSize = Math.min(this.size, (1 << (k - 1)) + 1);
+  if (newSize < 2) {
+    if (newSize < 1)  {
```
Comment (lemire) [inline FastPriorityQueue.js]: If you add the guard above, then newSize cannot be zero because this.size>0, and (1<<(k-1)+1) is at least 2. So this branch would not be needed.

You could just say `if(newSize == 1) { return [this.peek()]; }`.

File: FastPriorityQueue.js [hunk 4]
```diff
@@ -287,13 +287,22 @@ FastPriorityQueue.prototype.forEach = function(callback) {
...
+      return [];
+    } else if (k > 0) {
+      return [this.peek()];
+    }
+  }
+  if (k < 1) return [];
```
Comment (lemire) [inline FastPriorityQueue.js]: If you guard at the start for this, as I suggest, then this line is unneeded.

**Comment by lemire** [comment]
@Barbarrosa This is very good. Please consider my comments.

File: FastPriorityQueue.js (hunk 2 above)
Comment (Barbarrosa) [reply FastPriorityQueue.js]: I implemented this change after testing. Similar checks I had tested at this point degraded performance, but this did not.

File: FastPriorityQueue.js (hunk 3 above)
Comment (Barbarrosa) [reply FastPriorityQueue.js]: I found `newSize == 1` appeared to reduce performance, so I went with `newSize < 2`. This will be safe because `this.size` and `(1 << (k - 1)) + 1` are both guaranteed to be `>= 1` when this is checked (excepting cases that were also previously unsupported).

File: FastPriorityQueue.js (hunk 4 above)
Comment (Barbarrosa) [reply FastPriorityQueue.js]: `k < 1` check removed.

File: FastPriorityQueue.js (hunk 1 above)
Comment (Barbarrosa) [reply FastPriorityQueue.js:290]: Updated with the guard you suggested below.

**Comment by Barbarrosa** [comment]
@lemire I updated the branch based on your comments. Here is the updated performance based on my testing today.

Before (slow)

```
starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 25.80 ops/sec ±0.42% (46 runs sampled)

starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 2.70 ops/sec ±1.92% (11 runs sampled)
```

Original PR (faster)

```
starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 39.60 ops/sec ±0.95% (52 runs sampled)

starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 4.06 ops/sec ±0.80% (15 runs sampled)
```

PR after changes (similar to Original PR)

```
starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 39.81 ops/sec ±0.67% (53 runs sampled)

starting dynamic queue/enqueue benchmark
FastPriorityQueue--queue-and-kSmallest x 4.03 ops/sec ±1.56% (15 runs sampled)
```

**Comment by lemire** [comment]
Great. Merging.

**Comment by staltz** [comment]
Just a heads up (I don't have more information at the moment), this PR broke a corner case in [jitdb](https://github.com/ssbc/jitdb), something about calling `kSmallest` and it returning undefined. I'll try to work on finding a simple reproduction and submit a test to FastPriorityQueue. 

cc @arj03

**Comment by Barbarrosa** [comment]
@staltz I identified the issue, it's an overflow when requesting larger numbers of records. I'll open a PR to fix it.

**Comment by Barbarrosa** [comment]
@staltz @lemire I opened https://github.com/lemire/FastPriorityQueue.js/pull/38 to address the issue.
